File: core/MaximumProductSubarray/MaximumProductSubarray.py
from typing import List


def maxProduct(nums: List[int]) -> int:
    pre_min, pre_max = nums[0], nums[0]
    length, res = len(nums), nums[0]
    for i in range(1, length):
        pre_max, pre_min = max(max(nums[i] * pre_min, nums[i] * pre_max), nums[i]), min(min(nums[i] * pre_min, nums[i] * pre_max), nums[i])
        res = max(res, pre_max)

    return res

File: core/MaximumProductSubarray/MaximumProductSubarray.java
package com.dbc;

public class MaximumProductSubarray {
    public int maxProduct(int[] nums) {
        int min = nums[0], max = nums[0], res = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int temp_min = Math.min(Math.min(nums[i], nums[i] * min), nums[i] * max);
            int temp_max = Math.max(Math.max(nums[i], nums[i] * min), nums[i] * max);
            min = temp_min;
            max = temp_max;
            res = Math.max(res, max);
        }
        return res;
    }
}
